canals/pipeline/pipeline.py

```diff
@@ -344,24 +345,42 @@ def run(self, data: Dict[str, Any], debug: bool = False) -> Dict[str, Any]:
                 self._record_pipeline_step(step, inputs_buffer, pipeline_output)
             logger.debug("> Queue at step %s: %s", step, {k: list(v.keys()) for k, v in inputs_buffer.items()})
 
+            if len(inputs_buffer) > 0 and not [i for i in inputs_buffer.values() if i]:
+                # Everything in the inputs buffer is skipped, this either means we
+                # have a really nasty bug or that the Pipeline ran everything it could.
+                # Either way, bail!
+                break
+
             component, inputs = inputs_buffer.popitem(last=False)  # FIFO
 
+            # Make sure it didn't run too many times already
+            self._check_max_loops(component)
+
             # if debug:
             #     draw(deepcopy(self.graph), engine="graphviz", path=f"debug/step_{current_step}.jpg", running=component, queued=inputs_buffer.keys())
 
             # **** IS IT MY TURN YET? ****
-            # Check if the node should be run or not
-            ready_to_run = self._ready_to_run(name=component, inputs=inputs, inputs_buffer=inputs_buffer)
+            # Check if the component should be run or not
+            state = self._calculate_component_state(name=component, inputs=inputs, inputs_buffer=inputs_buffer)
```

Comment:
I don't like `_calculate_component_state` either to be fair. 

I would have very much prefered to go with three different methods `_should_run`, `_should_wait` and `_should_skip` that would return a boolean but that would have complicated the code a bit. Though it would have made it a bit easier to test too. I'm still debating with myself whether that's better or worse.

I decided to go with `_calculate_component_state` as it's clear that's returning the state of `component`. I find `_ready_to_run` to be semantically incorrect, I would expect a method  named like that to return a boolean but it doesn't. Same thing could be said for `_can_run_now` and `_expected_inputs_are_present`. 

That's why I decided to use `_calculate_component_state`, maybe we could change it to `_get_state`, `_component_state` or something similar. In any case I'd like it to be clear that it must not be used like the example below as that would be prone to errors.

```
if _component_state():
```